predicators/refinement_estimators/cnn_refinement_estimator.py

```diff
@@ -0,0 +1,138 @@
+    """A refinement cost estimator that uses a CNN to predict refinement cost
+    from an initial state render."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        # _model_dict maps immutable skeleton atoms_sequence pair to model
+        self._model_dict: Optional[Dict[ModelDictKey, CNNRegressor]] = None
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "cnn"
+
+    @property
+    def is_learning_based(self) -> bool:
+        return True
+
+    def get_cost(self, initial_task: Task, skeleton: List[_GroundNSRT],
+                 atoms_sequence: List[Set[GroundAtom]]) -> float:
+        assert self._model_dict is not None, "Need to train"
+        key = self._immutable_model_dict_key(skeleton, atoms_sequence)
+        # If key isn't in dictionary (no data for skeleton), cost is infinite
+        if key not in self._model_dict:
+            return float('inf')
+        model = self._model_dict[key]
+        # Get render image of initial state and pass into model
+        input_img = self._get_rendered_initial_state(initial_task)
+        cost = model.predict(input_img)
+        return cost[0]
+
+    def train(self, data: List[RefinementDatapoint]) -> None:
+        """Train the CNN regressors on the data points for that skeleton,
+        atoms_sequence pair."""
+
+        # Go through data and group them by skeleton
+        grouped_input_imgs = defaultdict(list)
+        grouped_targets = defaultdict(list)
+        for task, skeleton, atoms_sequence, succeeded, refinement_time in data:
+            # Convert skeleton and atoms_sequence into an immutable dict key
+            key = self._immutable_model_dict_key(skeleton, atoms_sequence)
+            # Render the initial state for use as an input image matrix
+            img = self._get_rendered_initial_state(task)
+            grouped_input_imgs[key].append(img)
+            # Compute target value from refinement time and possible failure
+            value = refinement_time
+            if not succeeded:
+                value += CFG.refinement_data_failed_refinement_penalty
+            grouped_targets[key].append([value])
+
+        # For each (skeleton, atoms_sequence) key, fit a CNNRegressor
+        self._model_dict = {}
+        total_num_keys = len(grouped_input_imgs)
+        for i, key in enumerate(grouped_input_imgs):
+            X = np.stack(grouped_input_imgs[key])
+            assert len(X.shape) == 4  # expect (N, 3, H, W)
+            Y = np.array(grouped_targets[key])
+            assert Y.shape == (X.shape[0], 1)
+            model = self._create_regressor()
+            logging.info(f"Training CNN for skeleton {i}/{total_num_keys} "
+                         f"using {X.shape[0]} data points...")
+            t0 = time.perf_counter()
+            model.fit(X, Y)
+            logging.info(f"Fit model in {time.perf_counter() - t0:.2f}s")
+            self._model_dict[key] = model
+
+    @staticmethod
+    def _immutable_model_dict_key(
```

Comment:
building on my question about unifying `CostDictKey` and `ModelDictKey`, I'm wondering if we can remove this copied code as well. how about implementing a parent class for this and the tabular approach called `PerSkeletonRefinementEstimator`?